src/libponyc/codegen/genident.c

```diff
@@ -284,21 +284,43 @@ static LLVMValueRef gen_is_value(compile_t* c, ast_t* left_type,
       l_value = LLVMBuildBitCast(c->builder, l_value, c->object_ptr, "");
       r_value = LLVMBuildBitCast(c->builder, r_value, c->object_ptr, "");
 
-      if(!is_known(left_type) && !is_known(right_type))
-      {
-        reach_type_t* r_left = reach_type(c->reach, left_type);
-        reach_type_t* r_right = reach_type(c->reach, right_type);
-        int possible_boxes = subtype_kind_overlap(r_left, r_right);
+      bool left_known = is_known(left_type);
+      bool right_known = is_known(right_type);
+      reach_type_t* r_left = reach_type(c->reach, left_type);
+      reach_type_t* r_right = reach_type(c->reach, right_type);
 
-        if((possible_boxes & SUBTYPE_KIND_BOXED) != 0)
-          return box_is_box(c, r_left, l_value, r_value, possible_boxes);
+      if(!left_known && !right_known)
+      {
+        int sub_kind = subtype_kind_overlap(r_left, r_right);
+
+        if((sub_kind & SUBTYPE_KIND_BOXED) != 0)
+          return box_is_box(c, r_left, l_value, r_value, sub_kind);
+
+        // If the types can be the same, check the address.
+        if(sub_kind != SUBTYPE_KIND_NONE)
+          return LLVMBuildICmp(c->builder, LLVMIntEQ, l_value, r_value, "");
+      } else if(left_known && right_known) {
```

Comment:
`subtype_kind_overlap` currently requires both types to be abstract types. Since the other part of the compiler using this function (`gen_is` in `genident.c`) currently doesn't need to handle known types through this function, I'm inclined to leave it as is to avoid paying a performance penalty on `gen_is`. That performance penalty would be rather small so I'm open to other arguments here.